test_util/provide_aws.py

```diff
@@ -0,0 +1,215 @@
+        self.stack = None
+        self.stack_name = None
+        self.region = region
+        self.session = boto3.session.Session(
+            aws_access_key_id=os.environ['AWS_ACCESS_KEY_ID'],
+            aws_secret_access_key=os.environ['AWS_SECRET_ACCESS_KEY'])
+        self.ec2_client = self.session.client(
+            service_name='ec2',
+            region_name=self.region,
+            verify=AWS_VERIFY)
+
+    def set_expiration(self, expiration_string):
+        ''' For use with Mesosphere/DCOS CI
+        Instances not tagged and owned after after instantiation will be deleted
+        within 30 minutes
+        '''
+        reservation = self.ec2_client.describe_instances(Filters=[{
+                'Name': 'tag-value', 'Values': [self.stack_name]}])['Reservations']
+        for inst_i in reservation[0]['Instances']:
+            inst = self.session.resource('ec2', region_name=self.region).Instance(inst_i['InstanceId'])
+            inst.create_tags(
+                    DryRun=False,
+                    Tags=[
+                        {
+                            'Key': 'expiration',
+                            'Value': expiration_string
+                        },
+                        {
+                            'Key': 'owner',
+                            'Value': '#dcos-ci'
+                        }])
+
+    def create_key_pair(self, key_name):
+        """
+        Args:
+            key_name (str): key name to be registered in AWS
+        Returns:
+            string of contents of key
+        """
+        key = self.ec2_client.create_key_pair(KeyName=key_name)
+        return key['KeyMaterial']
+
+    def delete_key_pair(self, key_name):
+        self.ec2_client.delete_key_pair(KeyName=key_name)
+
+    @retrying.retry(stop_max_delay=1200*1000,
+                    wait_fixed=3000,
+                    retry_on_exception=lambda ex: False,
+                    retry_on_result=lambda res: res is False)
+    def wait_for_cluster_creation(self):
+        logger.info('Waiting for cluster to be created...')
+        self.stack.reload()
+        current_status = self.stack.stack_status
+        if current_status == 'CREATE_IN_PROGRESS':
+            return False
+        if current_status != 'CREATE_COMPLETE':
+            raise Exception("Expected stack {} to have status CREATE_COMPLETE,"
+                            "but found: {}".format(self.stack.stack_id,
+                                                   current_status))
+
+    @retrying.retry(stop_max_delay=1200*1000,
+                    wait_fixed=3000,
+                    retry_on_exception=lambda ex: False,
+                    retry_on_result=lambda res: res is False)
+    def wait_for_cluster_deletion(self):
+        logging.info('Waiting for cluster to be deleted...')
+        current_status = self.stack.stack_status
+        if current_status == 'DELETE_IN_PROGRESS':
+            return False
+        try:
+            self.stack.reload()
```

Comment:
if we use the boto3 client directly (https://boto3.readthedocs.io/en/latest/reference/services/cloudformation.html#CloudFormation.Client.describe_stacks) we can set `StackName=<stack_id>` at which point the deleted stacks will still show up and can actually look for just DELETE_COMPLETE simplifying this code a bit.